src/peft/tuners/lora/eva.py

```diff
@@ -0,0 +1,463 @@
+    The rank distribution for each layer is determined based on the explained variance ratio.
+
+    Parameters:
+    -----------
+    model: (torch.nn.Module)
+        The model to compute the SVD for.
+    peft_config: (LoraConfig)
+        The configuration for the LoRA layers.
+    dataloader: (torch.utils.data.DataLoader)
+        The dataloader to use for the forward pass.
+    forward_fn: callable
+        The forward function to use for the forward pass.
+        `model(**inputs)` is used if forward_fn is not provided.
+    get_indices_fn: (Optional[callable])
+        The function to use if not all positions in the input tensor should be used for SVD 
+        (e.g. for causal language modeling). Can be set to None if all positions should be used.
+        `peft.tuners.lora.eva.get_indices_fn_causal_lm` is used by default.
+        Should always return a tensor of shape (n_indices, layer_input.ndim-1).
+    prepare_activations_fn: (Union[callable, Dict[str, callable], None])
+        If a layer recieves multiple inputs as a list, per default the first input is used.
+        This function can be used to modify this behaviour. Accepts a dictionary with layer names as keys.
+        Default logic:
+        ```
+        try:
+            svd_input = activations.detach()
+        except AttributeError:
+            svd_input = activations[0].detach()
+        ```
+        
+    Returns:
+    --------
+    eva_state_dict: (dict)
+        The state dictionary containing the SVD components for each layer.
+    """
+    
+    # Computes the rank distribution for each layer based on the explained variance ratio.
+    def _get_rank_distribution(hooks, hook_layer_map, equal_inputs_map, rank_budget, max_components):
+        exp_vars = {k: h.svd.explained_variance_ratio_[:max_components] for k, h in hooks.items()}
+        keys, values = zip(*[(k, c) for k, name in hook_layer_map.items() for c in exp_vars[name]])
+        idx = torch.stack(values).argsort(descending=True)
+        counts = Counter([keys[i] for i in idx[:rank_budget]])
+        counts = {k: counts.get(k, 0) for k in hook_layer_map.keys()} # add layers with 0 rank
+        for k, k_hook in equal_inputs_map.items():
+            # ensure hook layers have the highest rank if they are equal to another layer
+            rank, rank_hook = counts[k], counts[k_hook]
+            if rank_hook >= rank:
+                continue
+            counts[k_hook], counts[k] = rank, rank_hook
+        return counts
+
+    # set function to check if modules should be added to hooks
+    if hasattr(model, "peft_config"):
+        _check_fn = lambda name, module: isinstance(module, Linear)
+    else:
+        _check_fn = lambda name, module: any([name.endswith(t) for t in peft_config.target_modules])
+
+    training = model.training
+    device = get_device_with_meta_params(model)
+    model.eval()
+
+    hooks = {}
```

Comment:
Do we need a check at the end that all the hooks have been cleaned up or is that not necessary?

Reply:
good idea! I implemented a check now
```python
    # check all svd hooks have been removed
    remaining_svd_hooks = {n for n, m in model.named_modules() for v in m._forward_hooks.values() if isinstance(v, SVDHook)}
    if len(remaining_svd_hooks) > 0:
        raise ValueError(f"All SVD hooks should have been removed but the following layers still have active hooks: {remaining_svd_hooks}")
```